cmd/tsbs_generate_queries/databases/clickhouse/devops.go

```diff
@@ -0,0 +1,341 @@
+func (d *Devops) getHostWhereString(nhosts int) string {
+	hostnames := d.GetRandomHosts(nhosts)
+	return d.getHostWhereWithHostnames(hostnames)
+}
+
+// getSelectClausesAggMetrics gets specified aggregate function clause for multiple memtrics
+// Ex.: max(cpu_time) AS max_cpu_time
+func (d *Devops) getSelectClausesAggMetrics(aggregateFunction string, metrics []string) []string {
+	selectAggregateClauses := make([]string, len(metrics))
+	for i, metric := range metrics {
+		selectAggregateClauses[i] = fmt.Sprintf("%[1]s(%[2]s) AS %[1]s_%[2]s", aggregateFunction, metric)
+	}
+	return selectAggregateClauses
+}
+
+// ClickHouse understands and can compare time presented as strings of this format
+const clickhouseTimeStringFormat = "2006-01-02 15:04:05"
+
+// MaxAllCPU selects the MAX of all metrics under 'cpu' per hour for nhosts hosts,
+// e.g. in pseudo-SQL:
+//
+// SELECT MAX(metric1), ..., MAX(metricN)
+// FROM cpu
+// WHERE
+// 		(hostname = '$HOSTNAME_1' OR ... OR hostname = '$HOSTNAME_N')
+// 		AND time >= '$HOUR_START'
+// 		AND time < '$HOUR_END'
+// GROUP BY hour
+// ORDER BY hour
+//
+// Resultsets:
+// cpu-max-all-1
+// cpu-max-all-8
+func (d *Devops) MaxAllCPU(qi query.Query, nHosts int) {
+	interval := d.Interval.RandWindow(devops.MaxAllDuration)
+	metrics := devops.GetAllCPUMetrics()
+	selectClauses := d.getSelectClausesAggMetrics("max", metrics)
+
+	sql := fmt.Sprintf(`
+        SELECT
+            toStartOfHour(created_at) AS hour,
+            %s
+        FROM cpu
+        WHERE %s AND (created_at >= '%s') AND (created_at < '%s')
+        GROUP BY hour
+        ORDER BY hour
+        `,
+		strings.Join(selectClauses, ", "),
+		d.getHostWhereString(nHosts),
+		interval.Start.Format(clickhouseTimeStringFormat),
+		interval.End.Format(clickhouseTimeStringFormat))
+
+	humanLabel := devops.GetMaxAllLabel("ClickHouse", nHosts)
+	humanDesc := fmt.Sprintf("%s: %s", humanLabel, interval.StartString())
+	d.fillInQuery(qi, humanLabel, humanDesc, sql)
+}
+
+// GroupByTimeAndPrimaryTag selects the AVG of numMetrics metrics under 'cpu' per device per hour for a day,
+// e.g. in pseudo-SQL:
+//
+// SELECT AVG(metric1), ..., AVG(metricN)
+// FROM cpu
+// WHERE time >= '$HOUR_START' AND time < '$HOUR_END'
+// GROUP BY hour, hostname
+// ORDER BY hour
+//
+// Resultsets:
+// double-groupby-1
+// double-groupby-5
+// double-groupby-all
+func (d *Devops) GroupByTimeAndPrimaryTag(qi query.Query, numMetrics int) {
+	metrics := devops.GetCPUMetricsSlice(numMetrics)
+	interval := d.Interval.RandWindow(devops.DoubleGroupByDuration)
+
+	selectClauses := make([]string, numMetrics)
+	meanClauses := make([]string, numMetrics)
+	for i, m := range metrics {
+		meanClauses[i] = "mean_" + m
+		selectClauses[i] = fmt.Sprintf("avg(%s) AS %s", m, meanClauses[i])
+	}
+
+	hostnameField := "hostname"
+	joinClause := ""
+	if d.UseTags {
+		//hostnameField = "tags.hostname"
```

Comment:
Let's get rid of this comment

Reply:
Yes, this part of code is not clear and looks like some development retries - I do not remember clearly what was that, sorry. Anyway, need to be cleaned.
Updated